### DISCUSSION TITLE
Cannot use lfs (lua file system) in LUA filter plugin.

### DISCUSSION THREAD

**Discussion by maitry-nutanix** [fluent/fluent-bit#7749] [Q&A]
### PROBLEM
I need to find the size of a log file in lua script. I can do that by opening and reading the file (using the code below) but that doesn't seem like a good solution as this script runs for every log line read by fluenbit.
```
        local file_size = file:seek('end')
        file:close()
```
### CODE
I cannot use lfs in LUA script in LUA filter plugin. I used: 
```
local lfs = require "lfs"
local size = lfs.attributes (path, "size")
```
### ERROR
But getting an error as follows: 
```
[2023/07/25 13:44:43] [error] [filter:lua:lua.1] error code 2: /fluent-bit/etc/..2023_07_25_13_44_22.370465307/filters.lua:6: module 'lfs' not found:
	no field package.preload['lfs']
	no file './lfs.lua'
	no file '/usr/local/share/luajit-2.1.0-beta3/lfs.lua'
	no file '/usr/local/share/lua/5.1/lfs.lua'
	no file '/usr/local/share/lua/5.1/lfs/init.lua'
	no file './lfs.so'
	no file '/usr/local/lib/lua/5.1/lfs.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
```
### HELP NEEDED
1. Is there any other way to use lfs or it is not supported yet. 
2. Open for any other solutions instead of using lfs. Please share if you have any ideas.

**Comment by patrick-stephens** [comment]
LUA modules may need system dependencies so these need to be installed. It looks like you need to install the dependencies required.

**Reply by maitry-nutanix** [reply]
@patrick-stephens this will be done by creating a custom flb image which includes these dependencies right ?

**Reply by patrick-stephens** [reply]
Yes, if you're using container images.

Be aware the official images use Distroless so you have to install dependencies via multistage build. Refer to the Dockerfile to see how we do it for the current build.

**Reply by maitry-nutanix** [reply]
@patrick-stephens thanks for your help. I tried to create a new image using current Dockerfile of fluentbit. I was able to access lfs library also but the new image size is 1.32GB. 

Here are the steps I used: 
ran  `docker build -t my-fluentbit-image:version . `  
then created a tar file for the same using  `docker save -o my-fluentbit-image-version my-fluentbit-image:version`
then uploaded the tar on the filer and finally loaded the docker image using `docker load`. 

 Using the above steps the image created was of size 1.32GB
 `custom-fluentbit                                                              original                       75c6cb161256        13 minutes ago      1.32GB`
 
 Is there any other way to create an image ??

**Reply by patrick-stephens** [reply]
We build the image from the Dockerfile yes. I'm not really sure about the rest of the question...

I think what you're doing is building the debug image, the Dockerfile has multiple targets so will default to the debug one if you do not specify.

You can see how the actual releases are done here: https://github.com/fluent/fluent-bit/blob/cfca7ca8acdaca04229c64efb1f85206fc291a37/.github/workflows/call-build-images.yaml#L115-L131

For production images you need to specify the [`--target production`]: https://docs.docker.com/engine/reference/commandline/build/#target

Note that production images do not have a shell or package manager though so you will not be able to use them to add more dependencies.
As I indicated above, you need to look at the multi-stage approach we use in the Dockerfile to add more dependencies.

I would also say, you can just make your own Dockerfile using the OSS image as the base - there is no need to copy the whole Dockerfile.

**Reply by arunim2405** [reply]
@jadiym can you please share the dockerfile that you ended up with? We need to implement something similar where we need to use lua-zlib.
@patrick-stephens I also tried using a compile luac.out file as a script but that unfortunately did not work. Do you see it as a valid feature request ? (support for compiled lua files)

**Reply by patrick-stephens** [reply]
I don't think we support compiled lua files but depends on what LuaJIT does.

**Reply by tchrono** [reply]
@arunim2405 fluent-bit simply embeds LuaJIT engine. If LuaJIT supports loading precompiled modules (it should support), then it should work under fluent-bit.


Here's an article that might help you: https://blog.openresty.com/en/luajit-bytecode/

**Reply by tchrono** [reply]
BTW I'm not sure LuaJIT bytecode is portable across versions. If you do this, make sure to use the exact same version of LuaJIT used by fluent-bit.

**Reply by patrick-stephens** [reply]
It seems a brittle way to do it to be honest
